ompi/runtime/ompi_spc.h

```diff
@@ -0,0 +1,210 @@
+    OMPI_IGATHER,
+    OMPI_ALLTOALL,
+    OMPI_IALLTOALL,
+    OMPI_IALLTOALLV,
+    OMPI_IALLTOALLW,
+    OMPI_NEIGHBOR_ALLTOALL,
+    OMPI_NEIGHBOR_ALLTOALLV,
+    OMPI_NEIGHBOR_ALLTOALLW,
+    OMPI_ALLGATHER,
+    OMPI_ALLGATHERV,
+    OMPI_IALLGATHER,
+    OMPI_IALLGATHERV,
+    OMPI_NEIGHBOR_ALLGATHER,
+    OMPI_NEIGHBOR_ALLGATHERV,
+    OMPI_TEST,
+    OMPI_TESTALL,
+    OMPI_TESTANY,
+    OMPI_TESTSOME,
+    OMPI_WAIT,
+    OMPI_WAITALL,
+    OMPI_WAITANY,
+    OMPI_WAITSOME,
+    OMPI_BARRIER,
+    OMPI_IBARRIER,
+    OMPI_WTIME,
+    OMPI_CANCEL,
+    OMPI_BYTES_RECEIVED_USER,
+    OMPI_BYTES_RECEIVED_MPI,
+    OMPI_BYTES_SENT_USER,
+    OMPI_BYTES_SENT_MPI,
+    OMPI_BYTES_PUT,
+    OMPI_BYTES_GET,
+    OMPI_UNEXPECTED,
+    OMPI_OUT_OF_SEQUENCE,
+    OMPI_MATCH_TIME,
+    OMPI_UNEXPECTED_IN_QUEUE,
+    OMPI_OOS_IN_QUEUE,
+    OMPI_MAX_UNEXPECTED_IN_QUEUE,
+    OMPI_MAX_OOS_IN_QUEUE,
+    OMPI_SPC_NUM_COUNTERS /* This serves as the number of counters.  It must be last. */
+};
+
+/* There is currently no support for atomics on long long values so we will default to
+ * size_t for now until support for such atomics is implemented.
+ */
+typedef size_t ompi_spc_value_t;
+
+/* A structure for storing the event data */
+typedef struct ompi_spc_s{
+    char *name;
+    ompi_spc_value_t value;
+} ompi_spc_t;
+
+/* Events data structure initialization function */
+void events_init(void);
+
+/* OMPI SPC utility functions */
+void ompi_spc_init(void);
+void ompi_spc_fini(void);
+void ompi_spc_record(unsigned int event_id, ompi_spc_value_t value);
+void ompi_spc_timer_start(unsigned int event_id, opal_timer_t *cycles);
+void ompi_spc_timer_stop(unsigned int event_id, opal_timer_t *cycles);
+void ompi_spc_user_or_mpi(int tag, ompi_spc_value_t value, unsigned int user_enum, unsigned int mpi_enum);
+void ompi_spc_cycles_to_usecs(ompi_spc_value_t *cycles);
+void ompi_spc_update_watermark(unsigned int watermark_enum, unsigned int value_enum);
+static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, void *obj_handle);
+
+/* Macros for using the SPC utility functions throughout the codebase.
+ * If SPC_ENABLE is not 1, the macros become no-ops.
+ */
+#if SPC_ENABLE == 1
+
+#define SPC_INIT()  \
+    ompi_spc_init()
+
+#define SPC_FINI()  \
+    ompi_spc_fini()
+
+#define SPC_RECORD(event_id, value)  \
+    ompi_spc_record(event_id, value)
+
+#define SPC_TIMER_START(event_id, usec)  \
+    ompi_spc_timer_start(event_id, usec)
+
+#define SPC_TIMER_STOP(event_id, usec)  \
+    ompi_spc_timer_stop(event_id, usec)
+
+#define SPC_USER_OR_MPI(tag, value, enum_if_user, enum_if_mpi) \
+    ompi_spc_user_or_mpi(tag, value, enum_if_user, enum_if_mpi)
+
+#define SPC_CYCLES_TO_USECS(cycles) \
+    ompi_spc_cycles_to_usecs(cycles)
+
+#define SPC_UPDATE_WATERMARK(watermark_enum, value_enum) \
+    ompi_spc_update_watermark(watermark_enum, value_enum)
+
+#else /* SPCs are not enabled */
+
+#define SPC_INIT()  \
+    do {} while (0)
```

Comment:
_If_ this is the problem with XL, why don't you just have it:
```
#define SPC_INIT() ;
```

Reply:
I was under the impression that some compilers do not like empty statements such as ";;".

Reply:
hopefully this line is our problem.
We can test this hypothesis by putting something random there right?